### PR TITLE
add prettier code formatting and linting for import/export values

### PR DESCRIPTION
**Pull request by mwcz**
This adds prettier code formatting.  It will be applied when `eslint --fix` is run.  It also adds `eslint-plugin-import` which lints imports and exports.  For example if you try to `import { fooBlah } from "./lib.js"` and lib.js doens't have a fooBlah export, eslint will error.

It also adds husky and lint-staged to allow linting to happen when a file is staged in git. There was already a precommit npm script but precommit requires husky and (optionally) lint-staged to be installed in order to work.  npm itself doesn't have a precommit script.

### CODE REVIEW

**Comment by blalasaadri** [comment]
I would recommend adding the `examples` directory to the ESLint scan as well - turns out, there are quite a few issues there.
So, that would mean that in the `package.json` we would change the scripts definition to:
```json
"scripts": {
        "..."
        "eslint": "./node_modules/.bin/eslint src examples",
        "eslint-fix": "./node_modules/.bin/eslint --fix src examples",
        "..."
}
```
Also, I think missing JSDoc comments should be a warning rather than an error - any thoughts on this?

File: .eslintrc.json [hunk 1]
```diff
@@ -71,13 +88,8 @@
...
         ],
-        "no-unused-vars": [
-            "error"
-        ],
-        "brace-style": [
-            "error",
-            "stroustrup"
-        ],
+        "no-unused-vars": ["error"],
+        "brace-style": ["error", "stroustrup"],
```
Comment (blalasaadri) [inline .eslintrc.json]: This conflicts with `prettier`. If I have
```javascript
if (...) {
    // ...
} else {
    // ...
}
```
I get the error `error    Closing curly brace appears on the same line as the subsequent block  brace-style`. If however I switch to
```javascript
if (...) {
    // ...
}
else {
    // ...
}
```
I get `warning  Insert `⏎···················`                                         prettier/prettier`.

There has to be at least one correct way. Either we switch this to `1tbs` or we change the `prettier` settings to allow for the Stroustrup style.

File: .eslintrc.json [hunk 2]
```diff
@@ -32,26 +43,32 @@
...
-                    "exceptions": [
-                        "*"
-                    ],
+                    "exceptions": ["*"],
                     "balanced": true
                 }
             }
         ],
+        "require-jsdoc": [
+            "error",
```
[the remark this answers is not in the document]
Comment (mwcz) [reply .eslintrc.json]: Good catch, I thought I had removed that!

File: .eslintrc.json (hunk 1 above)
Comment (mwcz) [reply .eslintrc.json]: Awesome, I'll change this to `1tbs`.  I personally prefer stroustrup style but I think sticking to prettier's defaults is the right call.

**Comment by mwcz** [comment]
@blalasaadri I also wanted to add the examples dir, but it required cleanup I wasn't ready to do yet (time constraints).  Part of the problem is the `lib` dir, which has files we definitely don't want to lint.  I'll update this PR with a `.eslintignore` file to exclude `lib`.

**Comment by mwcz** [comment]
Thanks @blalasaadri !

**Comment by blalasaadri** [comment]
It seems the Travis build is now failing because it's merging this branch onto master and the `main_text.js` file there doesn't yet follow the new guidelines. I've looked into it and the only problem I can't seem to be able to fix is that the variable `text` is unused; maybe @JeffreyBrettColeman (who added that variable) can help here? My change with everything working except for that is at https://github.com/blalasaadri/hero-engine/commit/f43159358d824a35f566f13b715fbc9b649c68ec.

**Comment by mwcz** [comment]
@blalasaadri  @JeffreyBrettColeman  I applied @blalasaadri's fix and also removed the line length linting rule (prettier handles that for everything but comments, which don't seem worth enforcing).

File: .eslintrc.json [hunk 3]
```diff
@@ -1,27 +1,38 @@
...
+    },
+    "plugins": ["prettier"],
     "env": {
         "browser": true,
         "es6": true,
         "node": true
     },
-    "extends": [
-        "eslint:recommended",
-        "google"
```
[the remark this answers is not in the document]
Comment (mwcz) [reply .eslintrc.json]: We talked in person, but for everyone else: for "correctness" linting, like making sure variables are used, the eslint defaults are almost exactly what we want.  For coding style linting, prettier covers that.

File: .eslintrc.json [hunk 4]
```diff
@@ -32,26 +43,32 @@
...
+            {
+                "require": {
+                    "FunctionDeclaration": false,
+                    "MethodDefinition": false,
+                    "ClassDeclaration": false,
+                    "ArrowFunctionExpression": false,
+                    "FunctionExpression": false
+                }
+            }
+        ],
```
[the remark this answers is not in the document]
Comment (mwcz) [reply .eslintrc.json]: Good catch, that was a mistake.  I don't see it anymore so I guess one of us fixed it. :D

**Pull request merged**